### ISSUE TITLE
Webextension support

### ISSUE THREAD

**Issue by anthnub** [carlin-q-scott/browser-media-keys#150]
Not sure if Firefox webextensions APIs allow for an update, but would definitely appreciate it if possible

**Comment by svnpenn** [comment]
https://github.com/carlin-q-scott/browser-media-keys/issues/108

**Comment by svnpenn** [comment]
https://github.com/carlin-q-scott/browser-media-keys/issues/148

**Comment by svnpenn** [comment]
https://github.com/carlin-q-scott/browser-media-keys/issues/149

**Comment by bwydoogh** [comment]
Any update on this? Really missing this extension on Firefox Nightly.

**Comment by svnpenn** [comment]
https://github.com/carlin-q-scott/browser-media-players/issues/1

**Comment by mFoxRU** [comment]
Same here, missing it in Firefox Developer edition.

**Comment by jerome-tgl** [comment]
FYI, Firefox 57 will be released in the stable channel next week.

If I'm not mistaken, the 3,030 current Media Keys users won't be able to use it as it won't work anymore (cf. [legacy add-ons](https://support.mozilla.org/en-US/kb/firefox-add-technology-modernizing)), unless they switch to Firefox SRE which should drop support for legacy extension in the coming months.

**Comment by svnpenn** [comment]
@jerome-tgl the devs here cant really do anything about making this work until
Mozilla implements `MediaPlayPause` (which has no current ETA):

http://bugzilla.mozilla.org/show_bug.cgi?id=1251795

and `global` (which wont be fixed until at least FF58):

http://bugzilla.mozilla.org/show_bug.cgi?id=1411795

**Comment by svnpenn** [comment]
@jerome-tgl the devs here cant really do anything about making this work until
Mozilla implements `MediaPlayPause` (which has no current ETA):

http://bugzilla.mozilla.org/show_bug.cgi?id=1251795

and `global` (which wont be fixed until at least FF58):

http://bugzilla.mozilla.org/show_bug.cgi?id=1411795

The Mozilla devs @andymckay want to introduce WebExtentions to the community -
fine - but the APIs should be finished first

**Comment by svnpenn** [comment]
https://github.com/andymckay/arewewebextensionsyet.com/issues/20

**Comment by svnpenn** [comment]
https://github.com/andymckay/arewewebextensionsyet.com/issues/21

**Comment by svnpenn** [comment]
https://github.com/carlin-q-scott/browser-media-keys/issues/153

**Comment by svnpenn** [comment]
Firefox 57 is less than a week away, with [media keys][1251795] and
[global keys][1411795] still not implemented. Until this happens, you can stop
Firefox from automatically installing updates:

~~~js
Services.prefs.setBoolPref('app.update.auto', false);
~~~

[1411795]://bugzilla.mozilla.org/show_bug.cgi?id=1411795
[1251795]://bugzilla.mozilla.org/show_bug.cgi?id=1251795

**Comment by carlin-q-scott** [comment]
Hey, sorry everyone for not getting around to this.

I was able to port the website side code over to a WebExtension months ago but they (Mozilla) want me to write native console apps for each operating system (Mac, Linux and Windows) and that's such a complicated/messy task that I just haven't had the energy to dig into it. @Noitidart wrote the Mac/Linux support code and it's huge which makes those especially challenging. I'll try my best to get Windows support live this week but there's little chance of me completing Linux/Mac support anytime soon.

**Comment by svnpenn** [comment]
not sure what you are talking about. the only "app" is Firefox - and to extend it is to write an addon - of course any addons would not work outside the scope of Firefox, and so would not be called apps - so what you are saying is nonsensical

until the issues i linked above are fixed - it is currently impossible for this addon to work with FF57/webextention

**Comment by svnpenn** [comment]
@carlin-q-scott not sure what you are talking about. the only "app" is Firefox - and to extend it is to write an addon - of course any addons would not work outside the scope of Firefox, and so would not be called apps - so what you are saying is nonsensical

until the issues i linked above are fixed - it is currently impossible for this addon to work with FF57/webextention

**Comment by svnpenn** [comment]
@carlin-q-scott Native messaging:

http://developer.mozilla.org/Add-ons/WebExtensions/Native_messaging

this to me seems like an awful idea, and if this addon goes that way I will drop
use of it. the simple solution to get this addon working with FF57/webextension
is for mozilla to simply implement the required APIs:

- http://bugzilla.mozilla.org/show_bug.cgi?id=1251795
- http://bugzilla.mozilla.org/show_bug.cgi?id=1411795

as Chrome and Opera have already done:

http://developer.mozilla.org/Add-ons/WebExtensions/manifest.json/commands#Browser_compatibility

until then I am recommended people stop updates:

~~~js
Services.prefs.setBoolPref('app.update.auto', false);
~~~

**Comment by Noitidart** [comment]
> this to me seems like an awful idea, and if this addon goes that way I will drop
use of it. 

@svnpenn I agree with you. I personally do not use any native messaging addons (unless it was written by me), because the native binary is completely unreviewed and can do anything and no one will ever know. The only option if we don't want to go native messaging, is to wait for the WebExt system to get media key support.

**Comment by theromi** [comment]
@carlin-q-scott Is there any update on this? I'd prefer a Quantum-compatible version with possible concerns over it not working at all to be honest.

**Comment by svnpenn** [comment]
@theromi read my previous comment - this addon will continue to be broken until mozilla implements the required APIs

**Comment by svnpenn** [comment]
https://github.com/carlin-q-scott/browser-media-keys/issues/155

**Comment by theromi** [comment]
@svnpenn It seems to me that you were suggesting not to use Native Messaging, but that it'd still be possible?

**Comment by svnpenn** [comment]
@theromi you shouldnt be asking yourself "can we do this", you should be asking yourself "should we do this".

The answer is no. Native Messaging is a hack to account for the current crippled API.

**Comment by theromi** [comment]
@svnpenn That's why I was asking @carlin-q-scott for his opinion on this. I would **greatly** prefer a temporary hack over not upgrading to Quantum.

**Comment by svnpenn** [comment]
@theromi youre not listening to me. If you read the previous page I linked:

http://bugzilla.mozilla.org/show_bug.cgi?id=1251795

you will notice that @carlin-q-scott himself commented on the issue. He knows this addon is broken - and he knows, as I do, that nothing should be done about that until the proper APIs are implemented by mozilla. Stop pushing Native Messaging or anything else, because all it does is start the slippery slope. Accept that you need to use Firefox 56 for now, or that you cant use this addon until the API is implemented.

**Comment by theromi** [comment]
@svnpenn The comment you linked is 7 month old, while the comment by @carlin-q-scott in this thread is 10 days old - where he said that a version with Native Messaging for Windows is on the way. 

**Please stop making things up.** This isn't helping at all.

**Comment by svnpenn** [comment]
@theromi stop grasping at Native Messaging as a liferaft. It is not that. Accept
the fact that Mozilla messed up, and that you can either keep Firefox 56, or
move on and lose this addon for the time being. Native Messaging is not a
solution because:

1. its not portable (have to compile separate version for each OS)
2. its not javascript (violates web standards)
3. it encourages fragmentation between browsers - when the whole point of
   webextensions was to reduce that

stop pushing this for your own selfish reasons and realize the problem is bigger
than this one addon

**Comment by Noitidart** [comment]
Actually the biggest issue with native messaging is it is unreviewed. The only person that knows what the binary does is the author. So you have to trust the author is not doing anything malicious. Complete blackbox.

**Comment by svnpenn** [comment]
@theromi Id like to add - shouting here is like shouting into the wind, it might
make you feel better but it does nothing. If you actually want to do something,
comment on these issues here:

1. http://bugzilla.mozilla.org/show_bug.cgi?id=1251795
2. http://bugzilla.mozilla.org/show_bug.cgi?id=1411795
3. http://github.com/andymckay/arewewebextensionsyet.com/issues/20
4. http://github.com/andymckay/arewewebextensionsyet.com/issues/21

**Comment by theromi** [comment]
> Accept the fact that Mozilla messed up
I do accept that.

> realize the problem is bigger than this one addon
I do realizie that.

> stop pushing this for your own selfish reasons
I don't. You do.

> Native Messaging is not a solution
It is! **For now.** 

I do understand your concerns. I completly understand your reasoning on this. But the conclusions you draw are overly idealistic. _From my point of view_ Native Messaging is perfectly acceptable for the time being. Not choosing to go down this path is reactionary and quixotic.

Anyway, I don't think that our discussion will bring this any further. Let's just wait for @carlin-q-scott, he knows all the facts and is the one to decide.

> If you actually want to do something, comment on these issues here:
I will do that, too.

**Comment by svnpenn** [comment]
@theromi you are kidding yourself bro. You think @carlin-q-scott is going to be
your savior here, he is not. I am thankful to him for writing this addon, but
he realizes as I do that she should do nothing until the APIs are finished. Just
look at his activity:

http://github.com/carlin-q-scott

nothing in 4 months. Realize that you are not waiting on him. You are, as we all
are, waiting on Mozilla to write the APIs. If you dont like it, you can stick
with Firefox 56, or go to Google Chrome. Or you could write the Native Messaging
yourself, but that seems unlikely given your activity:

http://github.com/theromi

Cheers.

**Comment by theromi** [comment]
> but he realizes as I do that she should do nothing until the APIs are finished. Just look at his activity:
There is no causal connection between this two sentences.

> Or you could write the Native Messaging yourself, but that seems unlikely given your activity:
Again, I fail to see your point. You seem to be overly attached to commitment to an arbitrary website.

If you'd like to bring any valuable comments to this thread, please do. But please stop trying to support your cause with mindless falsehood or false ad hominem.

**Comment by Noitidart** [comment]
@theromi I think what @svnpenn is trying to ask is. Would you be comfortable installing an unreviewed blackboxed software "native messaging" that can be doing malicious stuff on your computer?

**Comment by svnpenn** [comment]
@theromi [dont be thick in front of me][1]. If someone has not made a single
contribution in over 4 months, probability dictates that behavior will continue.

Is it a causal relationship? No. but to assume he is just going to magically
pop out a Native Messaging app after 4 months of nothing is wishful thinking at
best. Stop it.

[1]://youtube.com/watch?v=rghDqZYTvvc&t=1m20s

**Comment by Noitidart** [comment]
> wishful thinking at best. Stop it.

@svnpenn - whats the problem with wishful thinking? People can hope if they want. That's kind of what we're doing with hoping Mozilla completes the APIs. Sure probabilities are different, but to each their own right?

**Comment by svnpenn** [comment]
@Noitidart [wrong. wrong.][1] This is a javascript repository, and pestering
@carlin-q-scott to write a C app to account for failure in Mozilla APIs is
essentially off topic, which is what @theromi is doing.

I am using Firefox 56, I have turned off updates. I have accepted the situation
as it currently is. I do not like it either, but I am being realistic about it.
If people want to hound someone, they should hound Mozilla, which @theromi has
not done:

- http://bugzilla.mozilla.org/show_bug.cgi?id=1251795
- http://bugzilla.mozilla.org/show_bug.cgi?id=1411795
- http://github.com/andymckay/arewewebextensionsyet.com/issues/20
- http://github.com/andymckay/arewewebextensionsyet.com/issues/21

[1]://youtube.com/watch?v=Mw8NR6p8gyI&t=2m54s

**Comment by Noitidart** [comment]
@svnpenn my question to you was:

>  whats the problem with wishful thinking? People can hope if they want.

You said

> @Noitidart wrong. wrong. 

So people can't hope if they want? They can't wish?

**Comment by svnpenn** [comment]
@Noitidart if you want to boil my 84 words down to 2, then I dont think I have anything further to say to you - good day

**Comment by Noitidart** [comment]
Well I asked you a question, so in those words I was looking for the answer. I didn't think "This is a javascript respository", "writing a C app", "Firefox 56", "external links" was an answer to "People can hope if they want?". Only thing that applied was your "wrong wrong".

**Comment by theromi** [comment]
> Would you be comfortable installing an unreviewed blackboxed software "native messaging" that can be doing malicious stuff on your computer?

Good argument. You are right, I am not comfortable doing this. Unfortunately, this is already the case as a part-time Windows-User. While, of course, I'd like to keep this in check, this is one of the cases I'd support going in that direction (temporarily). From my point of view, having a suboptimal solution in place is a lot better than not moving on at all.

> If someone has not made a single contribution in over 4 months, probability dictates that behavior will continue.

Commitment to "leisure projects" are different for everyone, please keep that in mind.

> but to assume he is just going to magically pop out a Native Messaging app after 4 months of nothing is wishful thinking

No. That is just quite normal.

> This is a javascript repository, and pestering @carlin-q-scott to write a C app to account for failure in Mozilla APIs is essentially off topic, which is what @theromi is doing.

**I am not pestering anyone here.** I just asked for a status on the already announced update. Which is when you came in and started irrationally attacking anyone in sight.

> I am using Firefox 56, I have turned off updates. I have accepted the situation as it currently is. 

If you want to mark time: Okay. That doesn't mean that the world around you has to adapt to your delicacy.

> I do not like it either, but I am being realistic about it.

As I already elongated explained, to which you couldn't bring up any sensible counter-arguments: You are the one being unrealistic. Fortunately, the world works different and is not treading water.

> if you want to boil my 84 words down to 2, then I dont think I have anything further to say to you - good day

I think it's quite funny that you bring up this argument, while boiling down all comments you answer to single, mostly untrue summaries while making false assupmtions that fit your narrative. Your style of reasoning here is an embarrassment.

**Comment by svnpenn** [comment]
> From my point of view, having a suboptimal solution in place is a lot better
> than not moving on at all.

Of course you would say this, **you are a help vampire**:

http://meta.stackoverflow.com/a/252077

You contribute nothing:

http://github.com/theromi

You dont realize what goes into writing a C app, I do:

http://github.com/svnpenn

> I am not pestering anyone here

Yeah, you are. This issue has been posted:

#108

and posted:

#148
#149

and posted again:

#150
#153
#155

but somehow you thought it was a good idea to remind @carlin-q-scott, again,
that this issue is not fixed. **he knows**. Good day.

**Comment by svnpenn** [comment]
> From my point of view, having a suboptimal solution in place is a lot better
> than not moving on at all.

@theromi of course you would say this, **you are a help vampire**:

http://meta.stackoverflow.com/a/252077

You contribute nothing:

http://github.com/theromi

You dont realize what goes into writing a C app, I do:

http://github.com/svnpenn

> I am not pestering anyone here

Yeah, you are. This issue has been posted:

#108

and posted:

#148
#149

and posted again:

#150
#153
#155

but somehow you thought it was a good idea to remind @carlin-q-scott, again,
that this issue is not fixed. **he knows**. Good day.

**Comment by theromi** [comment]
> of course you would say this, you are a help vampire:
> You contribute nothing:

This is wrong. Just making this assumption based on a Github profile - which I mostly have for commenting and issue reporting - is false and foolish. _As already mentioned._ 

> You dont realize what goes into writing a C app, I do:

I contributed to more C "apps" than your profile even shows - so I am the winner here? Based on your reasoning, we should stick to what I say now! (sic)
Believe it or not, writing C code is even part of my job! _As already mentioned_, I am completly capable of understanding your sparse arguments. I am, though, not capable of tolerating your untenable ad homiem attacks. You are damaging your reputation. Which is probably why you deleted most of your posts.

> but somehow you thought it was a good idea to remind @carlin-q-scott,

And again you seem to ignore anything that has been said. _As already said,_ I was merely asking for a status update, _which is not given in any of the threads you linked_. I **was not** reminding anyone of anything he or she has to do.

As you now have repeatedly shown, you are incapable of disproving valid arguments (ignoring them again and again), you are incapable of keeping this discussion on a dignified level and you are makting up "facts", I will assume that you are are not keen on keeping this project sound. The quality and style of your posts leave no other conclusion than that you are a troll. And I should have noticed that a few posts ago.

**Comment by svnpenn** [comment]
@theromi all a bunch of talk - you provide no evidence - not a single link or reference to back up your claims - I will give your response the respect is deserves - none.

**Comment by Quentin1010** [comment]
Can someone move/remove those chitchats (mine included) out to keep the issue clean please ? This is an issue about some software, not some battlefield for people who don't seem to understand when an argument is useless...

**Comment by svnpenn** [comment]
@Quentin1010 I have removed mine - except the one actually germane to the issue - thanks

**Comment by svnpenn** [comment]
Currently the Firefox [commands API][compat] does not support
[media keys][1251795] nor [global keys][1411795].

Until these issues are fixed by Mozilla, you can use [Chromium][chrom] with the
[Streamkeys extension][stream].

[1251795]://bugzilla.mozilla.org/show_bug.cgi?id=1251795
[1411795]://bugzilla.mozilla.org/show_bug.cgi?id=1411795
[chrom]://github.com/henrypp/chromium
[compat]://developer.mozilla.org/Add-ons/WebExtensions/manifest.json/commands#Browser_compatibility
[stream]://www.streamkeys.com

**Comment by svnpenn** [comment]
Actually, it appears that media keys **are working**, just not globally. Example
"manifest.json":

~~~json
{
  "name": "wh",
  "version": "1.0.0",
  "manifest_version": 2,
  "background": {
    "scripts": ["xr.js"]
  },
  "commands": {
    "ya": {
      "suggested_key": {
        "default": "MediaPlayPause"
      }
    }
  }
}
~~~

and "xr.js":

~~~js
browser.commands.onCommand.addListener(
  zu => console.log(zu)
);
~~~

Until global is implemented:

http://bugzilla.mozilla.org/show_bug.cgi?id=1411795

the above will only work if the browser has focus, but **it does work**. If you
need global, you can use Chromium with the Streamkeys extension:

- http://github.com/henrypp/chromium
- http://www.streamkeys.com

**Comment by svnpenn** [comment]
Here is a proof of concept - WebExtension that supports media keys:

http://github.com/svnpenn/bm/tree/master/w-youtube-pp

**Comment by ldavison** [comment]
Sorry if this is off topic, but @svnpenn I wasn't able to get your PoC to work on Firefox Nightly 59.0a1 (2017-12-12) (64-bit) on macOS 10.12.6.

To simplify the extension, I created a new directory with the two below files and ran `web-ext run -f nightly`. The only command that worked was Ctrl+Shift+U, none of the media keys commands worked for me. Am I missing something?

<details>
<summary>manifest.json</summary>
<pre><code>
{
  "name": "command-test",
  "version": "1.0.0",
  "manifest_version": 2,
  "background": {
    "scripts": ["background.js"]
  },
  "commands": {
    "toggle-feature": {
      "suggested_key": {
        "default": "Ctrl+Shift+U"
      }
    },
    "media-toggle-playback": {
      "suggested_key": {
        "default": "MediaPlayPause"
      }
    },
    "media-prev-track": {
      "suggested_key": {
        "default": "MediaPrevTrack"
      }
    },
    "media-next-track": {
      "suggested_key": {
        "default": "MediaNextTrack"
      }
    }
  }
}
</code></pre>
</details>

<details>
<summary>background.js</summary>
<pre><code>
'use strict';
browser.commands.getAll().then((all) => all.forEach((c) => console.log("Registered command:", c)));
browser.commands.onCommand.addListener((c) => console.log("Received command:", c));
</code></pre>
</details>

**Comment by svnpenn** [comment]
@ldavison post here and i will look at it

http://github.com/svnpenn/umber/issues

**Comment by svnpenn** [comment]
https://github.com/svnpenn/umber/issues/27

**Comment by dvdbot** [comment]
I'm curious what's the progress on the update to Firefox Quantum?
Used the add-on a lot before, really handy to skip a song in a playlist by just using the mediakeys on my headset

**Comment by svnpenn** [comment]
@dvdbot media keys **are working**, just not globally. Here is a proof of concept;
WebExtension that supports media keys:

http://github.com/svnpenn/umber/tree/master/webext/raspberry

Until global is implemented:

http://bugzilla.mozilla.org/show_bug.cgi?id=1411795

the above will only work if the browser has focus, but it does work. If you
need global, you can use Chromium with the Streamkeys extension:

- http://github.com/henrypp/chromium
- http://www.streamkeys.com

**Comment by cup** [comment]
https://github.com/carlin-q-scott/browser-media-keys/issues/158

**Comment by escaton** [comment]
@carlin-q-scott what do you think about releasing new version compatible with FF but with no global listener feature?

**Comment by carlin-q-scott** [comment]
@escaton That's easy to do but I'm surprised people would be interested in that.

**Comment by escaton** [comment]
In my opinion half working extension is better then nothing, especially if it's easy as you said.

**Comment by CennoxX** [comment]
> 
> @escaton That's easy to do but I'm surprised people would be interested in that.

I would really like such an extension! For the time being it's also easy to write a small external script (with something like AutoHotkey) that always passes media keys to Firefox.

**Comment by bastien30** [comment]
Same for me, I would love an extension like that !

**Comment by carlin-q-scott** [comment]
Ok, I will release the WebExtension as-is since it already does this. It'll take a little while for it to go through the approval process though.

**Comment by carlin-q-scott** [comment]
I'm sorry to say that I have to eat my words. Firefox currently only supports the PlayPause media key and it doesn't support global media keys. So you need to have the media tab active and can only pause or resume playing. Opera and Chrome work fine.
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/commands#Browser_compatibility

**Comment by carlin-q-scott** [comment]
Ok, it's published. https://addons.mozilla.org/en-US/firefox/addon/media-keys/

Due to the conversion, some features have been removed or modified:
1. The browser window must be active for media keys to be detected.
2. Only Play/Pause is supported; this is a Firefox limitation; see https://bugzilla.mozilla.org/show_bug.cgi?id=1251795
3. Pinned tabs get priority over all other tabs when pressing a media key.
4. The autoplay feature is disabled by default and cannot be enabled until I create a configuration page for the add-on.

**Comment by cup** [comment]
Note that Version 2.0.0 of this addon does not work with Firefox 56. Make sure
to turn Automatic Updates off.

**Comment by carlin-q-scott** [comment]
@cup I think I fixed that automatic update issue by marking 2.0 as requiring Firefox 57+. Can you check to see if that's true?

**Comment by cup** [comment]
@carlin-q-scott looks good - thanks

hopefully Mozilla will get their act together one day :(

**Comment by leo60228** [comment]
@carlin-q-scott Where's the source code of the new version? I'd like to add Hulu support.
